### PR TITLE
Add SGMCMC module and corresponding bayesian_nn example

### PR DESCRIPTION
**Pull request by csy530216**
Features of SGMCMC module:
- Include SGLD, PSGLD (implemented by @meta-inf), SGHMC and SGNHT.
- Support 2nd-order local integrator introduced in [this paper](https://arxiv.org/pdf/1610.06665.pdf).

Two examples:
- Toy model: Sample from 1-D mixture of two Gaussians, illustrating that SGNHT, ideally, can correct the unknown noise brought by discretization and stochastic gradient.
- Bayesian NN for regression: On UCI Protein dataset. The network weights are sampled through SGHMC, and the prior stds of weights are updated using Monte Carlo EM algorithm.

### CODE REVIEW

**Comment by thjashin** [comment]
Is it close to the state that is fine to merge? If so, we can organize a meeting to decide on all the stuffs and finish it.

**Comment by thjashin** [comment]
py2.7 failed.

**Comment by thjashin** [comment]
need to merge from v4? I've pushed some fixes in tests.

File: examples/bayesian_neural_nets/bayesian_nn_sgmcmc.py [hunk 1]
```diff
@@ -0,0 +1,139 @@
...
+    # sgmcmc = zs.SGLD(learning_rate=1e-5, add_noise=True)
+    sgmcmc = zs.SGHMC(learning_rate=2e-6, friction=0.2, n_iter_resample_v=1000,
+                      second_order=True)
+    # sgmcmc = zs.SGNHT(learning_rate=1e-5, variance_extra=0., tune_rate=50.,
+    #                   second_order=True)
+    latent = dict(zip(w_names, wv))
+
+    # E step: Sample the parameters
+    sgmcmc.make_grad_func(meta_bn, observed={'x': x, 'y': y}, latent=latent)
+    sample_op, new_w, sample_info = sgmcmc.sample()
```
Comment (thjashin) [inline examples/bayesian_neural_nets/bayesian_nn_sgmcmc.py]: Like HMC, here is no need to return new_w?

File: zhusuan/sgmcmc.py [hunk 2]
```diff
@@ -0,0 +1,546 @@
...
+    instead of the true gradients. To deal with the problems brought by
+    stochasticity in gradients, more sophisticated updating scheme, such as
+    SGHMC and SGNHT, were proposed. We provided four SGMCMC algorithms here:
+    SGLD, PSGLD, SGHMC and SGNHT. For SGHMC and SGNHT, we support 2nd-order
+    integrators introduced in (Chen et al., 2015).
+
+    The implementation framework is similar to that of
+    :class:`~zhusuan.hmc.HMC` class. However, SGMCMC algorithms do not include
+    Metropolis update, and typically do not include hyperparameter adaptation,
+    so the implementation here is easier to read than that of
```
Comment (thjashin) [inline zhusuan/sgmcmc.py]: remove "so the implementation .."?

File: zhusuan/sgmcmc.py [hunk 3]
```diff
@@ -0,0 +1,546 @@
...
+            samples.
+
+        :return: A default gradient function for sampling algorithm, whose
+            argument is a list of tensorflow Variables corresponding to the
+            latent variables.
+        :return: The same gradient function whose arguments are the `Variable`
+            list above **and** a `observed` dictionary, to allow users to
+            create a gradient function with dataset splitting in case of large
+            batch size and limited memory.
+        :return: The `Variable` list which will serve as the argument of the
```
Comment (thjashin) [inline zhusuan/sgmcmc.py]: why are these variables needed?

File: zhusuan/sgmcmc.py [hunk 4]
```diff
@@ -0,0 +1,546 @@
...
+            return tf.gradients(
+                _get_log_posterior(var_list, observed), var_list)
+
+        self._default_get_gradient = \
+            lambda var_list: _get_gradient(var_list, observed)
+        self._latent_k = latent_k
+        self._var_list = latent_v
+        return self._default_get_gradient, _get_gradient, latent_v
+
+    def sample(self, grad_func=None):
```
Comment (thjashin) [inline zhusuan/sgmcmc.py]: Let's rename this to ``apply_updates`` and add another ``sample`` that does ``compute_grad_func`` as well as ``apply_updates``.

File: zhusuan/sgmcmc.py [hunk 5]
```diff
@@ -0,0 +1,546 @@
...
+
+class SGLD(SGMCMC):
+    """
+    Subclass of SGMCMC which implements Stochastic Gradient Langevin Dynamics
+    (Welling & Teh, 2011) (SGLD) update. The updating equation implemented
+    below follows Equation (3) in the paper.
+
+    :param learning_rate: A 0-D `float32` Tensor. It can be either a constant
+        or a placeholder for decaying learning rate.
+    :param add_noise: A `Bool` Tensor, if set to False, then the algorithm runs
```
Comment (thjashin) [inline zhusuan/sgmcmc.py]: may not need ``add_noise``? also for PSGLD?
